File: src-tauri/src/commands/mentions.rs
use crate::commands::network::NetworkState;
use crate::{
    error::AppError,
    services::{
        MentionReceipt, MentionsService, PublishMentionedPostRequest, PublishMentionedPostResult,
        ResolvedMention,
    },
};
use std::sync::Arc;
use tauri::State;

#[tauri::command]
pub async fn resolve_private_mention(
    service: State<'_, Arc<MentionsService>>,
    network: State<'_, NetworkState>,
    qualified_name: String,
) -> Result<ResolvedMention, AppError> {
    let resolved = service.resolve(&qualified_name)?;
    if resolved.status == "unknown" {
        let handle = network.get_handle().await?;
        let relay = handle.active_relay().await?;
        let (key, expires) = handle
            .resolve_delivery_key(relay, qualified_name.clone())
            .await?;
        service.cache_delivery_key(&qualified_name, key, expires)?;
    }
    service.resolve(&qualified_name)
}
#[tauri::command]
pub fn create_post_with_mentions(
    service: State<'_, Arc<MentionsService>>,
    request: PublishMentionedPostRequest,
) -> Result<PublishMentionedPostResult, AppError> {
    service.publish(request)
}
#[tauri::command]
pub fn list_pending_mentions(
    service: State<'_, Arc<MentionsService>>,
) -> Result<Vec<MentionReceipt>, AppError> {
    service.pending()
}
#[tauri::command]
pub fn review_private_mention(
    service: State<'_, Arc<MentionsService>>,
    mention_id: String,
    decision: String,
) -> Result<(), AppError> {
    service.review(&mention_id, &decision)
}

File: src/services/mentions.ts
import type {
  MentionReceipt,
  PublishMentionedPostRequest,
  PublishMentionedPostResult,
  ResolvedMention,
} from '../types';
import { invokeCommand } from './command';

export const mentionsService = {
  resolve(qualifiedName: string): Promise<ResolvedMention> {
    return invokeCommand('resolve_private_mention', { qualifiedName });
  },
  publish(request: PublishMentionedPostRequest): Promise<PublishMentionedPostResult> {
    return invokeCommand('create_post_with_mentions', { request });
  },
  listPending(): Promise<MentionReceipt[]> {
    return invokeCommand('list_pending_mentions');
  },
  review(
    mentionId: string,
    decision: 'accept-notification' | 'accept-repost' | 'decline' | 'block',
  ): Promise<void> {
    return invokeCommand('review_private_mention', { mentionId, decision });
  },
};
